README.md
Retinotopy Analysis
===

This toolbox allows you to analyze data from a standard phase-encoded fMRI paradigm. Functions are mainly based on a tutorial written by Sam Schwarzkopf. The tutorial can be found [here](https://figshare.com/articles/journal_contribution/Retinotopy_tutorial/1206288).

## Installation
To use the functions of this toolbox, include the following code into your `startup.m` file:

```matlab
%------------- Retinotopy ----------------------------%
path_spm12 = <PATH-TO-SPM12-ROOT>;
path_retinotopy = <PATH-TO-RETINOTOPY-ROOT>;

% add spm12 and reinotopy functions
addpath(genpath(path_spm12));
addpath(fullfile(path_retinotopy, 'retinotopy'));

clear path_spm12 path_retinotopy;
%-----------------------------------------------------%
```

You have to specify the paths to the root directory of this toolbox and to the directory of your [SPM12](https://www.fil.ion.ucl.ac.uk/spm/software/spm12/) installation.

External functions are copyrighted by their respective authors and might not be covered under the same GPL license as the rest of this toolbox.

## Usage
Two example scripts can be found in the `scripts` folder.


CHANGELOG.md
# Changelog
History of major code changes:

## 2022-10-19
Add phase_style parameter to generate polar angle data that matches the convention in Noah Benson's [neuropythy](https://github.com/noahbenson/neuropythy) library.


scripts/multipol.m
% Multipol analysis
%
% This script runs a retinotopy analysis for computing the BOLD amplitude
% from a phase-encoding paradigm. The analysis includes the following 
% steps: (1) baseline correction, (2) percent signal change (psc) 
% conversion, (3) Fourier analysis, (4) output of metrics at stimulus
% frequency. For baseline correction, a time series with prefix b is 
% written. For psc conversion, a time series with prefix p is written. If 
% the number of cycles <freq> is not an integer, the first cycle fraction 
% is discarded from analysis. Cleanup will delete all generated time series 
% at the end.

% input data
input.data = {
    '/data/pt_01880/test/run/uadata.nii',...
    }; % anticlock
input.tr = 3; % repetition time in s
input.period = 48; % cycle period in s
input.fix = 12; % pre and post run baseline block in s
input.freq = 10.5; % number of cycles
input.hp_cutoff = 144; % cutoff frequency 1/cutoff in Hz
input.psc_cutoff = 50; % threshold value for psc conversion
input.phase_style = 'standard'; % standard | visual

% output specification
name_sess = 'SE_EPI2';
path_output = '/data/pt_01880/test';
cleanup = false;

%%% do not edit below %%%

for i = 1:length(input.data)

    % path and filename
    [path, file, ext] = fileparts(input.data{i});
    
    % run baseline correction
    dh_baseline_correction(...
        input.data{i},...
        input.tr,...
        input.hp_cutoff);

    % run psc conversion
    dh_psc_conversion(...
        fullfile(path,['b' file ext]),...
        input.psc_cutoff);
    
    % run frequency analysis
    dh_retino_fourier(...
        fullfile(path, ['pb' file ext]),...
        input.freq,...
        input.fix,...
        input.period,...
        input.tr);

    % run multipol phase
    dh_multipol_phase(...
        fullfile(path,['rpb' file '_real' ext]),...
        fullfile(path,['rpb' file '_imag' ext]),...
        input.freq,...
        name_sess,...
        path_output, ...
        input.phase_style);
    
    if cleanup
        delete(fullfile(path,['b' file ext]));
        delete(fullfile(path,['pb' file ext]));
        delete(fullfile(path,['rpb' file '_real' ext]));
        delete(fullfile(path,['rpb' file '_imag' ext]));
    end
    
end


scripts/retinotopy.m
% Retinotopy analysis
%
% This script runs a retinotopy analysis for a phase encoding paradigm
% with rings and wedges. The analysis includes the following steps: 
% (1) baseline correction, (2) percent signal change (psc) conversion, 
% (3) phase calculation, (4) averaging over sessions with opposite stimulus 
% direction. For baseline correction, a time series with prefix b is
% written. For psc conversion, a time series with prefix p is written. Note 
% that the number of cycles <freq> is not an integer value in my runs. The 
% first cycle fraction is discarded from further analysis. We use this to 
% have already initialised activation in the first volume.

% input data
input.pol.data.pos = '/data/pt_01880/test/pol_anticlock/uadata.nii'; % anticlock
input.pol.data.neg = '/data/pt_01880/test/pol_clock/uadata.nii'; % clock
input.pol.tr = 2; % repetition time in s
input.pol.period = 64; % cycle period in s
input.pol.fix = 12; % pre and post run baseline block in s
input.pol.freq = 8.25; % number of cycles
input.pol.hp_cutoff = 192; % cutoff frequency 1/cutoff in Hz
input.pol.psc_cutoff = 50; % threshold value for psc conversion
input.pol.phase_style = 'visual'; % standard | visual

input.ecc.data.pos = '/data/pt_01880/test/ecc_expanding/uadata.nii'; % expanding
input.ecc.data.neg = '/data/pt_01880/test/ecc_contracting/uadata.nii'; % contracting
input.ecc.tr = 2;
input.ecc.period = 32;
input.ecc.fix = 12;
input.ecc.freq = 8.25;
input.ecc.hp_cutoff = 96;
input.ecc.psc_cutoff = 50;
input.ecc.phase_style = 'standard'; % standard | visual

%%% do not edit below %%%

% concatenate all runs into single cell
all_data = {input.pol.data.pos input.pol.data.neg input.ecc.data.pos input.ecc.data.neg};
all_tr = [input.pol.tr input.pol.tr input.ecc.tr input.ecc.tr];
all_hp_cutoff = [input.pol.hp_cutoff input.pol.hp_cutoff input.ecc.hp_cutoff input.ecc.hp_cutoff];
all_psc_cutoff = [input.pol.psc_cutoff input.pol.psc_cutoff input.ecc.psc_cutoff input.ecc.psc_cutoff];
all_freq = [input.pol.freq input.pol.freq input.ecc.freq input.ecc.freq];
all_fix = [input.pol.fix input.pol.fix input.ecc.fix input.ecc.fix];
all_period = [input.pol.period input.pol.period input.ecc.period input.ecc.period];

% prepare path and filename
path = cell(1,length(all_data));
file = cell(1,length(all_data));
ext = cell(1,length(all_data));
for i = 1:length(all_data)
    [path{i}, file{i}, ext{i}] = fileparts(all_data{i});
end

% run baseline correction
for i = 1:length(all_data)
    dh_baseline_correction(...
        all_data{i},...
        all_tr(i),...
        all_hp_cutoff(i));
end

% run psc conversion
for i = 1:length(all_data)
    dh_psc_conversion(...
        fullfile(path{i},['b' file{i} ext{i}]),...
        all_psc_cutoff(i));
end

% run frequency analysis
for i = 1:length(all_data)
    dh_retino_fourier(...
        fullfile(path{i}, ['pb' file{i} ext{i}]),...
        all_freq(i),...
        all_fix(i),...
        all_period(i),...
        all_tr(i));
end

% run phase averaging
[pos_path, pos_file, pos_ext] = fileparts(input.pol.data.pos);
[neg_path, neg_file, neg_ext] = fileparts(input.pol.data.neg);
dh_average_phase(...
    fullfile(pos_path,['rpb' pos_file '_real' pos_ext]),...
    fullfile(pos_path,['rpb' pos_file '_imag' pos_ext]),...
    fullfile(neg_path,['rpb' neg_file '_real' neg_ext]),...
    fullfile(neg_path,['rpb' neg_file '_imag' neg_ext]),...  
    'pol',...
    input.pol.freq, ...
    input.pol.phase_style);

[pos_path, pos_file, pos_ext] = fileparts(input.ecc.data.pos);
[neg_path, neg_file, neg_ext] = fileparts(input.ecc.data.neg);
dh_average_phase(...
    fullfile(pos_path,['rpb' pos_file '_real' pos_ext]),...
    fullfile(pos_path,['rpb' pos_file '_imag' pos_ext]),...
    fullfile(neg_path,['rpb' neg_file '_real' neg_ext]),...
    fullfile(neg_path,['rpb' neg_file '_imag' neg_ext]),...  
    'ecc',...
    input.ecc.freq, ...
    input.ecc.phase_style);


retinotopy/dh_color_wheel.m
function dh_color_wheel(dir_out, radius)
% Color wheel
%
% dh_color_wheel(dir_out, radius)
%
% Inputs:
%   dir_our - Output directory.
%   radius  - Wheel radius in px.
%
% This function creates the polar angle RYGB colour wheel for freesurfer. 
% The script is adapted from Sam Schwarzkopf's examples in the retinotopy 
% tutorial.
%__________________________________________________________________________
% Copyright (C) 2021 Daniel Haenelt

if ~exist('radius','var')  
    radius = 200;
end

% colour peaks
R = [1 0 0];
Y = [1 1 0];
G = [0 1 0];
B = [0 0 1];

% make folder
if ~exist(dir_out,'dir') 
    mkdir(dir_out); 
end

% output image matrix
[x,y] = meshgrid(-radius:radius, -radius:radius);
[t,r] = cart2pol(x,y);
t = t / pi * 180; % convert to degrees
t = t + 90; % calibrate angles to stimulus starting position
t = mod(ceil(t),360) + 1; % ensure between 1-360
nr = ceil(r/radius*360); % normalized rho in degrees
nr(nr==0) = 1;

% polar angle
cycles = 2;
steps = 360 / 4 / cycles;
cmap = [];
for c = 1:cycles  
    cmap = [cmap; linspace(R(1), Y(1), steps)', linspace(R(2), Y(2), steps)', linspace(R(3), Y(3), steps)'];
    cmap = [cmap; linspace(Y(1), G(1), steps)', linspace(Y(2), G(2), steps)', linspace(Y(3), G(3), steps)'];
    cmap = [cmap; linspace(G(1), B(1), steps)', linspace(G(2), B(2), steps)', linspace(G(3), B(3), steps)'];
    cmap = [cmap; linspace(B(1), R(1), steps)', linspace(B(2), R(2), steps)', linspace(B(3), R(3), steps)'];
end

% image for output
imgR = zeros(2*radius+1, 2*radius+1);
imgG = zeros(2*radius+1, 2*radius+1);
imgB = zeros(2*radius+1, 2*radius+1);

imgR(r<radius) = cmap(t(r<radius),1);
imgG(r<radius) = cmap(t(r<radius),2);
imgB(r<radius) = cmap(t(r<radius),3);

% final image matrix
img = imgR;
img(:,:,2) = imgG;
img(:,:,3) = imgB;

% Flip left & right side (only affects pol)
img = flip(img, 2);

% copy right side and override left side
img_temp = flip(img, 2);
img(:,1:radius,:) = img_temp(:,1:radius,:);

% save image
imwrite(img, fullfile(dir_out, 'pol_legend.png'));

% eccentricity
cycles = 1;
steps = 360 / 3 / cycles;
cmap = [];
for c = 1:cycles  
    cmap = [cmap; linspace(R(1), Y(1), steps)', linspace(R(2), Y(2), steps)', linspace(R(3), Y(3), steps)'];
    cmap = [cmap; linspace(Y(1), G(1), steps)', linspace(Y(2), G(2), steps)', linspace(Y(3), G(3), steps)'];
    cmap = [cmap; linspace(G(1), B(1), steps)', linspace(G(2), B(2), steps)', linspace(G(3), B(3), steps)'];
end

% image for output
imgR = zeros(2*radius+1, 2*radius+1);
imgG = zeros(2*radius+1, 2*radius+1);
imgB = zeros(2*radius+1, 2*radius+1);

imgR(r<radius) = cmap(nr(r<radius),1);
imgG(r<radius) = cmap(nr(r<radius),2);
imgB(r<radius) = cmap(nr(r<radius),3);

% final image matrix
img = imgR;
img(:,:,2) = imgG;
img(:,:,3) = imgB;

% save image
imwrite(img, fullfile(dir_out, 'ecc_legend.png'));


retinotopy/dh_average_phase.m
function dh_average_phase(pos_real, pos_imag, neg_real, neg_imag, ...
    name_sess, freq, phase_style)
% Average phase
%
% dh_average_phase(pos_real, pos_imag, neg_real, neg_imag, ...
%    name_sess, freq, phase_style)
%
% Inputs:
%   pos_real    - real part of fft in positive direction.
%   pos_imag    - imaginary part of fft in positive direction.
%   neg_real    - real part of fft in negative direction.
%   neg_imag    - imaginary part of fft in negative direction.
%   name_sess   - name of session.
%   freq        - number of cycles.
%   phase_style - retinotopy styles (standard, visual).
%
% This function takes the real and imaginary parts from the time series fft
% of stimuli pairs in positive and negative directions. Both directions are
% averaged to compensate for the hemodynamic lag. Output files are saved in
% a created output folder with prefix s. Phase maps can be encoded by two
% different styles (standard, visual). The polar angle map with style
% 'visual' thereby matches the convention in Noah Benson's neuropythy 
% library.
%
% Standard:
%   left hemisphere: -90->-180,+180->90 (UVM -> LHM -> LVM)
%   right hemisphere: -90->+90 (UVM -> RHM -> LVM)
% Visual:
%   left hemisphere: 0->180 (UVM -> RHM -> LVM)
%   right hemisphere: 0->-180 (UVM -> LHM -> LVM)
%__________________________________________________________________________
% Copyright (C) 2021 Daniel Haenelt

if ~exist('phase_style', 'var')
    phase_style = 'standard';
end

phase_style_list = {'standard', 'visual'};
if ~any(strcmp(phase_style_list, phase_style))
    error('Error. Unknown phase_style!');
end

% parameters
% freq_ignored is a vector containing the frequencies that should be
% ignored when calculating the coherence values.
freq = floor(freq);
freq_ignored = [0:1 freq-1 freq freq+1];

% output folder is taken from the first entry of the function
path_output = fullfile(fileparts(fileparts(pos_real)),'avg','native');
if ~exist(path_output,'dir') 
    mkdir(path_output); 
end

% load input
data_img = spm_vol(neg_imag); % imaginary
c_imag = spm_read_vols(data_img);
data_img = spm_vol(neg_real); % real
c_real = spm_read_vols(data_img);
data_img = spm_vol(pos_imag); % imaginary
a_imag = spm_read_vols(data_img);
data_img = spm_vol(pos_real); % real
a_real = spm_read_vols(data_img);

% flip imaginary part of negative direction
c_imag = -c_imag;

% get image dimension
dim = data_img.dim;
nt = length(data_img);

% average different directions
all_real = NaN([dim nt 2]);
all_imag = NaN([dim nt 2]);
all_real(:,:,:,:,1) = c_real;
all_real(:,:,:,:,2) = a_real;
all_imag(:,:,:,:,1) = c_imag;
all_imag(:,:,:,:,2) = a_imag;
avg_real = mean(all_real,5);
avg_imag = mean(all_imag,5);

% mean power over all (considered) frequencies
avg_pw = sqrt(avg_real.^2+avg_imag.^2); 
mfs = 1:round(nt/2);
mfs = mfs(~ismember(mfs, freq_ignored+1));
mpw = nanmean(avg_pw(:,:,:,mfs),4); 
mstd = nanstd(avg_pw(:,:,:,mfs),0,4);

% real and imaginary parts at stimulus frequency
avg_imag_freq = squeeze(avg_imag(:,:,:,freq+1));
avg_real_freq = squeeze(avg_real(:,:,:,freq+1));   

% data at cycle frequency
avg_A = sqrt(avg_real_freq.^2+avg_imag_freq.^2); % amplitude
avg_F = avg_A ./ mpw * 100; % F-statistic 
avg_snr = avg_A ./ mstd; % SNR

% transform to phases
if strcmp(phase_style, 'standard')
    avg_pha = atan2(avg_imag_freq,avg_real_freq)/pi*180; % phase in degrees
end

if strcmp(phase_style, 'visual')
    avg_pha = (atan2(-1,0) - atan2(avg_imag_freq,avg_real_freq))/pi*180;
end

avg_pha = mod(avg_pha,360); % phases between 0 and 360
avg_pha = avg_pha - 180;  % phases between -180 and +180

% change NaN in background to 0
avg_real_freq(isnan(avg_real_freq)) = 0;
avg_imag_freq(isnan(avg_imag_freq)) = 0;
avg_pha(isnan(avg_pha)) = 0;
avg_A(isnan(avg_A)) = 0;
avg_F(isnan(avg_F)) = 0;
avg_snr(isnan(avg_snr)) = 0;

% save niftis
nhdr = data_img(1);
nhdr.fname = fullfile(path_output,[name_sess '_real_avg.nii']);
spm_write_vol(nhdr,avg_real_freq);

nhdr.fname = fullfile(path_output,[name_sess '_imag_avg.nii']);
spm_write_vol(nhdr,avg_imag_freq);

nhdr.fname = fullfile(path_output,[name_sess '_phase_avg.nii']);
spm_write_vol(nhdr,avg_pha);

nhdr.fname = fullfile(path_output,[name_sess '_a_avg.nii']);
spm_write_vol(nhdr,avg_A);

nhdr.fname = fullfile(path_output,[name_sess '_f_avg.nii']);
spm_write_vol(nhdr,avg_F);

nhdr.fname = fullfile(path_output,[name_sess '_snr_avg.nii']);
spm_write_vol(nhdr,avg_snr);


retinotopy/dh_psc_conversion.m
function dh_psc_conversion(input, cutoff_psc, prefix)
% Percent signal change conversion
%
% dh_psc_conversion(input, cutoff_psc, prefix)
%
% Inputs:
%   input      - file name of time series.
%   cutoff_psc - threshold value.
%   prefix     - prefix of output filename.
%
% This function converts a time series in arbitrary units into percent 
% signal change by dividing each voxel time series by its temporal mean and 
% multiplying it by 100. Optionally, all deviations larger than a set value 
% are thresholded. The output time series gets a prefix p to the file name.
%__________________________________________________________________________
% Copyright (C) 2021 Daniel Haenelt

if ~exist('cutoff_psc', 'var')
    cutoff_psc = 50;
end

if ~exist('prefix','var')  
    prefix = 'p';
end

% get fileparts of input
[path, file, ext] = fileparts(input);

% load input time series
data_img = spm_vol(input);
data_array = spm_read_vols(data_img);

% get image dimensions
dim = data_img(1).dim;
nt = length(data_img);

% compute percent signal change
data_array_mean = mean(data_array, 4);
data_array = data_array ./ data_array_mean * 100;

% threshold data
if cutoff_psc
    data_array_max = max(data_array, [], 4);
    data_array_min = min(data_array, [], 4);
    
    data_array_max = repmat(data_array_max, 1, 1, 1, nt);
    data_array_min = repmat(data_array_min, 1, 1, 1, nt);
    
    data_array(data_array_max > 100 + cutoff_psc) = 100 + cutoff_psc;
    data_array(data_array_min < 100 - cutoff_psc) = 100 - cutoff_psc;
end

% write output
for i = 1:nt
  data_img(i).dim = dim;
  data_img(i).fname = fullfile(path, [prefix file ext]);
  spm_write_vol(data_img(i), data_array(:,:,:,i));
end

retinotopy/dh_retino_fourier.m
function dh_retino_fourier(input, freq, fix, period, TR)
% Retino fourier
%
% dh_retino_fourier(input, freq, fix, period, TR)
%
% Inputs:
%   input  - file name of baseline corrected time series.
%   freq   - number of cycles.
%   fix    - pre and post run baseline block in s.
%   period - cycle period in s
%   TR     - repetition time in s.
%
% This function computes a fourier analysis for retinotopical data. Two
% time series are saved, namely the real and imaginary part of the Fourier
% transform. The data is thresholded by only including voxels above the 
% mean intensity. This is a very crude method for non-brain voxel
% exclusion. The output data gets the prefix r.
%__________________________________________________________________________
% Copyright (C) 2021 Daniel Haenelt

% prepare path and file name
[path, file, ext] = fileparts(input);

% number of dropped volumes
drop_start = round(fix/TR) + round(mod(freq,1)*period/TR);
drop_end = round(fix/TR);
drop = sum([drop_start drop_end]);

% get time series header
data_img = spm_vol(input);

% number of volumes
nt = length(data_img);

% load time series
data_array = spm_read_vols(data_img);
data_array = data_array(:,:,:,drop_start+1:nt-drop_end);

% shift time to first dim in time series matrix
ts = shiftdim(data_array,3);

% normalise time series
vm = repmat(mean(ts,1), [nt - drop 1 1 1]); % voxel-wise mean
ts = (ts - vm) ./ vm * 100; % voxel-wise percent signal change

% fourier transform
ft = fft(ts); % Fourier transform

% phase shift to cf (-90 deg)
% we start the stimulus from the 3 o'clock position. However, the first
% quarter of the period is discarded. Therefore, we apply a phase shift to
% compensate for that
ft = ft * exp(-1i*mod(freq,1)*2*pi);

% Real and imaginary parts
cI = -imag(ft); % imaginary component (negative because we want cosine as 0)
cR = real(ft); % real component

% Mask the brain
mask = data_array(:,:,:,1);
threshold = mean(mask(:)); % intensity threshold
nbvox = find(mask < threshold); % non-brain voxels
% Remove non-brain voxels
for i = 1:size(cI,1)
    tempI = cI(i,:,:,:);
    tempI(nbvox) = NaN;
    tempR = cR(i,:,:,:);
    tempR(nbvox) = NaN;
    cI(i,:,:,:) = tempI;
    cR(i,:,:,:) = tempR;
end

% shift time to last dimension
cI = shiftdim(cI,1);
cR = shiftdim(cR,1);

% write output
ndata_img = data_img(1:size(cI,4));
for i = 1:size(cI,4)
    ndata_img(i).fname = fullfile(path, ['r' file '_imag' ext]);
    ndata_img(i).pinfo = [1 0 0]';
    ndata_img(i).dt(1) = 16; % convert data type to double (float32)
    spm_write_vol(ndata_img(i),cI(:,:,:,i));
end
disp(['Saved imaginary image: ' ndata_img(1).fname]);

ndata_img = data_img(1:size(cR,4));
for i = 1:size(cR,4)
    ndata_img(i).fname = fullfile(path, ['r' file '_real' ext]);
    ndata_img(i).pinfo = [1 0 0]';
    ndata_img(i).dt(1) = 16; % convert data type to double (float32)
    spm_write_vol(ndata_img(i),cR(:,:,:,i));
end
disp(['Saved real image: ' ndata_img(1).fname]);


retinotopy/dh_baseline_correction.m
function dh_baseline_correction(input, TR, cutoff_highpass, prefix)
% Baseline correction
%
% dh_baseline_correction(input, TR, cutoff_highpass, prefix)
%
% Inputs:
%   input           - file name of time series.
%   TR              - repetition time in s.
%   cutoff_highpass - highpass 1/cutoff frequency in Hz.
%   prefix          - prefix of output filename.
%
% This function computes a baseline correction of a functional time series.
% The output time series gets a prefix b to the file name.
%__________________________________________________________________________
% Copyright (C) 2021 Daniel Haenelt

if ~exist('prefix','var')  
    prefix = 'b';
end

% get fileparts of input
[path, file, ext] = fileparts(input);

% load input time series
data_img = spm_vol(input);
data_array = spm_read_vols(data_img);

% get image dimensions
dim = data_img(1).dim;
nt = length(data_img);

% high-pass filtering of each voxel time-course
K.RT = TR;
K.row = 1:1:nt;
K.HParam = cutoff_highpass;

nK = spm_filter(K); % now we compute the low frequencies that should be removed from signal change
% nK.X0 contains the frequencies that can be removed. A second call of spm_filter removes these frequencies

data_array_corr = zeros([dim nt]);
for x = 1:dim(1)
    for y = 1:dim(2)
        for z = 1:dim(3)
            sig = data_array(x,y,z,:);
            sig = reshape(sig,[nt 1]);
            sig = spm_filter(nK,sig);
            sig = reshape(sig,[1 1 1 nt]);
            data_array_corr(x,y,z,:) = sig;
        end
    end
end

% write output
for i = 1:nt
  data_img(i).dim = dim;
  data_img(i).fname = fullfile(path, [prefix file ext]);
  spm_write_vol(data_img(i), data_array_corr(:,:,:,i));
end


retinotopy/dh_multipol_phase.m
function dh_multipol_phase(input_real, input_imag, freq, name_sess, ...
    path_output, phase_style)
% Multipol phase
%
% dh_multipol_phase(input_real, input_imag, freq, name_sess, ...
%    path_output)
%
% Inputs:
%   input_real  - real part of fft.
%   input_imag  - imaginary part of.
%   freq        - number of cycles.
%   name_sess   - name of session included in the basename.
%   path_output - path where output is saved.
%   phase_style - retinotopy styles (standard, visual).
%
% This function takes the real and imaginary parts from the time series
% fft and computes several metrics. Similar to Dumoulin et al. 2017, we 
% compute a coherence estimate of the voxel respose at stimulus frequency 
% by taking the amplitude and dividing it by the root mean square of the 
% one-sided power spectrum. Other measures as F-statistics are computed by 
% taking the amplitude and dividing it by the mean amplitude of the rest of 
% the frequency spectrum. Based on this, the voxel population can be 
% thresholded to investigate different subpopulations. To run this 
% function, the number of runs has to be specified to choose an appropriate 
% results folder location. Phase maps can be encoded by two different 
% styles (standard, visual). The polar angle map with style 'visual' 
% thereby matches the convention in Noah Benson's neuropythy library.
%
% Standard:
%   left hemisphere: -90->-180,+180->90 (UVM -> LHM -> LVM)
%   right hemisphere: -90->+90 (UVM -> RHM -> LVM)
% Visual:
%   left hemisphere: 0->180 (UVM -> RHM -> LVM)
%   right hemisphere: 0->-180 (UVM -> LHM -> LVM)
%__________________________________________________________________________
% Copyright (C) 2021 Daniel Haenelt

if ~exist('phase_style', 'var')
    phase_style = 'standard';
end

phase_style_list = {'standard', 'visual'};
if ~any(strcmp(phase_style_list, phase_style))
    error('Error. Unknown phase_style!');
end

% parameters
% freq_ignored is a vector containing the frequencies that should be
% ignored when calculating the denominator.
freq = floor(freq);
freq_ignored = [];
%freq_ignored = [0:1 freq-1 freq freq+1];

% define output paths
[~, name_run, ~] = fileparts(fileparts(input_real));
path_a = fullfile(path_output,'results','a','native');
path_c = fullfile(path_output,'results','c','native');
path_f = fullfile(path_output,'results','f','native');
path_imag = fullfile(path_output,'results','imag','native');
path_real = fullfile(path_output,'results','real','native');
path_phase = fullfile(path_output,'results','phase','native');
path_snr = fullfile(path_output,'results','snr','native');

if ~exist(path_a,'dir') 
    mkdir(path_a); 
end

if ~exist(path_c,'dir') 
    mkdir(path_c); 
end

if ~exist(path_f,'dir') 
    mkdir(path_f); 
end

if ~exist(path_imag,'dir') 
    mkdir(path_imag); 
end

if ~exist(path_real,'dir') 
    mkdir(path_real); 
end

if ~exist(path_phase,'dir') 
    mkdir(path_phase); 
end

if ~exist(path_snr,'dir') 
    mkdir(path_snr); 
end

% load input
data_img = spm_vol(input_imag); % imaginary
data_imag_array = spm_read_vols(data_img);
data_img = spm_vol(input_real); % real
data_real_array = spm_read_vols(data_img);

% get image dimension
nt = length(data_img);

% mean power over all (considered) frequencies
pw = sqrt(data_real_array.^2+data_imag_array.^2); 
mfs = 1:round(nt/2);
mfs = mfs(~ismember(mfs, freq_ignored+1));
sumpw = sqrt(nansum(pw(:,:,:,mfs).^2,4));
mpw = nanmean(pw(:,:,:,mfs),4); 
mstd = nanstd(pw(:,:,:,mfs),0,4);

% real and imaginary parts at stimulus frequency
data_imag_freq_array = squeeze(data_imag_array(:,:,:,freq+1));
data_real_freq_array = squeeze(data_real_array(:,:,:,freq+1));   

% data at cycle frequency
A_array = sqrt(data_real_freq_array.^2+data_imag_freq_array.^2); % amplitude
F_array = A_array ./ mpw * 100; % F-statistic 
snr_array = A_array ./ mstd; % SNR
coherence_array = A_array ./ sumpw; % coherence

% transform to phases
if strcmp(phase_style, 'standard')
    pha_array = atan2(data_imag_freq_array,data_real_freq_array)/pi*180; % phase in degrees
end

if strcmp(phase_style, 'visual')
    pha_array = (atan2(-1,0) - atan2(data_imag_freq_array,data_real_freq_array))/pi*180;
end

pha_array = mod(pha_array,360); % phases between 0 and 360
pha_array = pha_array - 180;  % phases between -180 and +180

% change NaN in background to 0
data_real_freq_array(isnan(data_real_freq_array)) = 0;
data_imag_freq_array(isnan(data_imag_freq_array)) = 0;
pha_array(isnan(pha_array)) = 0;
A_array(isnan(A_array)) = 0;
F_array(isnan(F_array)) = 0;
snr_array(isnan(snr_array)) = 0;
coherence_array(isnan(coherence_array)) = 0;

% save niftis
nhdr = data_img(1);

if isempty(name_sess)
    nhdr.fname = fullfile(path_real,['real_' name_run '.nii']);
    spm_write_vol(nhdr,data_real_freq_array);

    nhdr.fname = fullfile(path_imag,['imag_' name_run '.nii']);
    spm_write_vol(nhdr,data_imag_freq_array);

    nhdr.fname = fullfile(path_phase,['phase_' name_run '.nii']);
    spm_write_vol(nhdr,pha_array);

    nhdr.fname = fullfile(path_a,['a_' name_run '.nii']);
    spm_write_vol(nhdr,A_array);

    nhdr.fname = fullfile(path_f,['f_' name_run '.nii']);
    spm_write_vol(nhdr,F_array);

    nhdr.fname = fullfile(path_snr,['snr_' name_run '.nii']);
    spm_write_vol(nhdr,snr_array);

    nhdr.fname = fullfile(path_c,['c_' name_run '.nii']);
    spm_write_vol(nhdr,coherence_array);
else
    nhdr.fname = fullfile(path_real,['real_' name_sess '_' name_run '.nii']);
    spm_write_vol(nhdr,data_real_freq_array);

    nhdr.fname = fullfile(path_imag,['imag_' name_sess '_' name_run '.nii']);
    spm_write_vol(nhdr,data_imag_freq_array);

    nhdr.fname = fullfile(path_phase,['phase_' name_sess '_' name_run '.nii']);
    spm_write_vol(nhdr,pha_array);

    nhdr.fname = fullfile(path_a,['a_' name_sess '_' name_run '.nii']);
    spm_write_vol(nhdr,A_array);

    nhdr.fname = fullfile(path_f,['f_' name_sess '_' name_run '.nii']);
    spm_write_vol(nhdr,F_array);

    nhdr.fname = fullfile(path_snr,['snr_' name_sess '_' name_run '.nii']);
    spm_write_vol(nhdr,snr_array);

    nhdr.fname = fullfile(path_c,['c_' name_sess '_' name_run '.nii']);
    spm_write_vol(nhdr,coherence_array);
end
